File: rover.py
#!/usr/bin/python
'''
    A mars rover moves on a grid of fixed height and width according to a set of
    predefined rotation or movement instructions.
    
    We read an initial grid size as an integer, where the first digit describes the
    grid width, and the second digit describes the height.
    
    The following input line contains the starting position as an integer, followed 
    by the initial direction.
    
    The third and last input line contains a set of movement instructions, where:
        M = move one block in current direction
        L = rotate direction left
        R = rotate direction right
    
    This program outputs the final position (x, y) and direction the rover is facing.
'''
from sys import stdin, stdout, stderr
import re

class Rover(object):
    directions = "NESW"  # clockwise increment direction = 0,1,2,3 for N,E,S,W
    
    def __init__(self):
        ''' Read, interpret instructions, and report a final position after 
            applying instruction list
        '''
        try:
            width, height = self.grid_size()
            xy_pos, direction = self.initial_pos(width, height)
            movements = self.read_movements()        
            summary, direction = self.translate_movements(movements, direction)
            final_x = self.process_summary(summary, 3, 1, xy_pos[0], width)            
            final_y = self.process_summary(summary, 2, 0, xy_pos[1], height)
            if final_x < 0:
                raise ValueError('Horizontal movement out of range.\n')
            elif final_y < 0:
                raise ValueError('Vertical movement out of range.\n')
            else:
                stdout.write("%s %s %s\n" % (final_x, final_y, self.directions[direction]))            
        except ValueError as e:
            stderr.write(str(e))

    def grid_size(self):
        ''' Read the grid size from stdin. '''
        sz = re.match(r"^([1-9])([1-9])$", stdin.readline())
        if sz is None:
            msg = ('Invalid grid size provided. \n'
                   'Expected format is a two digit number with the first numeral\n'
                   ' as width, and the second numeral being height. eg. 88.\n')
            raise ValueError(msg)
        return int(sz.group(1)), int(sz.group(2))

    def initial_pos(self, width, height):
        ''' Read the initial position from stdin. '''
        init_pos = re.match(r"^([1-9])([1-9]) ([NSEW])$", stdin.readline())
        if init_pos is None:
            msg = ('Unexpected format for starting position and direction.\n'
                   'Expect: xy D\n'
                   ' where x and y are single digit numbers, and D is one of NESW.')
            raise ValueError(msg)
        xy_pos = (int(init_pos.group(1)), int(init_pos.group(2)))
        if xy_pos[0] > width or xy_pos[1] > height:
            msg = 'Initial coordinate positions exceed grid size.'
            raise ValueError(msg)
        direction = self.directions.index(init_pos.group(3))  # initial direction
        return xy_pos, direction

    def read_movements(self):
        ''' Read movements from stdin '''
        movements = stdin.readline().strip()
        if len(movements) is 0:
            raise ValueError('No movement instructions were supplied.\n')
        if len(set(movements) - set('LRM')) > 0:
            raise ValueError('Movement instructions may contain only L, R or M.\n')
            
        movements = list(movements)
        movements.reverse()   # set up movements stack
        return movements
    
    def process_summary(self, summary, minus, plus, pos, a_max):
        ''' Update current position using movement summaries,
            and check for bounds of the grid being exceeded.
        '''
        for direction, delta in summary:
            if direction == minus:
                pos -= delta 
            elif direction == plus:
                pos += delta
            if pos < 1 or pos > a_max:
                return -1
        return pos
        
    def turn(self, direction, lr):
        '''  Turn left or right by incrementing or decrementing
            the current direction.
        '''
        if lr == 'L':
            direction -= 1
        elif lr == 'R':
            direction += 1
        return (direction + 4) % 4

    def translate_movements(self, movements, direction):
        '''  Produce a summary of movements and a 
            final direction by stepping once through
            movement instructions.
        '''
        summary = []
        n = 0
        curr = None
        while len(movements):
            curr = movements.pop()
            if curr in ['L', 'R']:
                if n > 0:
                    summary.append((direction, n))
                direction = self.turn(direction, curr)
                n = 0
            elif curr == 'M':
                n += 1
        if n > 0:
            summary.append((direction, n))
        return summary, direction
        
if __name__ == "__main__":
    Rover()

File: rover.pl
% Uses SWI Prolog
%
%  Goal :-
%    Read the first line, interpret it as width and height,
%    Read the second line, interpret it as x, y pos and direction,
%    Read the third line, apply instructions in the line,
%    Output the final position and direction
%
%  With Prolog development, each predicate may succeed or fail, and
%  so testing and conditions form part of the code.
%
%  Each individual predicate may be tested in isolation.  For example,
%     turn(From, Dir, To),
%     writef("Turn %w from %w faces %w\n", [Dir, From, To]),
%     fail.
%  produces:
%     Turn L from E faces N
%     Turn L from S faces E
%     Turn L from W faces S
%     Turn L from N faces W
%     Turn R from N faces E
%     Turn R from E faces S
%     Turn R from S faces W
%     Turn R from W faces N
%
%  To run this program, install SWI Prolog, and do:
%     swipl -g script -s rover.pl
%
go :-
	get_gridsize(H, V),
	get_start_pos(H, V, X, Y, Direction),
	get_instructions(IList),
	process(pos(X, Y, Direction), size(H, V), IList, pos(Xf, Yf, Facing)),
	writef("%w %w %w", [Xf, Yf, Facing]), nl.

script :- go, fail.
script :- halt.

read_line(Line) :-  % Just read a single line and return the list of characters
	read_line_to_string(current_input, String),
	string_chars(String, Line).

% Reads the grid size as a two digit integer.  Fails if we cannot.
get_gridsize(H, V) :-
	read_line([Hc, Vc]),
	number_chars(H, [Hc]), number_chars(V, [Vc]),
	check_grid_range(H, V), !.
get_gridsize(_, _) :-
	write('Cannot read grid size.'), nl, fail.

% Checks the range of the grid size values
check_grid_range(H, V) :- between(1, 9, H), between(1, 9, V), !.
check_grid_range(H, V) :- writef('Grid values %w are out of range.', [[H, V]]), nl, fail.

% Reads the starting position and direction as a two digit integer
% followed by a space and a direction indicator.  Fails if it cannot.
get_start_pos(HSize, VSize, X, Y, Direction) :-
	read_line([Xc, Yc, ' ', Direction]),
	number_chars(X, [Xc]), number_chars(Y, [Yc]),
	check_grid_pos(X, HSize, 'Horizontal'), check_grid_pos(Y, VSize, 'Vertical'),
	check_direction(Direction), !.

% Checks that the position is valid, and fails if it is not.
check_grid_pos(Pos, Size, _) :- between(1, Size, Pos), !.
check_grid_pos(_, _, What) :- writef("%s position is outside the grid", [What]), nl, fail.

% Checks that the direction is in the valid set of directions, or fails
check_direction(Direction) :- memberchk(Direction, ['N', 'E', 'S', 'W']), !.
check_direction(Direction) :- writef("%s is not a valid direction", [Direction]), nl, fail.

% Reads the instruction list and checks that it is valid.
get_instructions(IList) :- read_line(IList), check_instructions(IList), !.

% Check whether each instruction in the list is valid, or fails.
valid_instruction_set([]).
valid_instruction_set([I|Iset]) :-
	writef("Instructions contain invalid characters: %w", [I|Iset]), nl, fail.
	
check_instructions([I|IList]) :-
	list_to_ord_set([I|IList], Iset),
	ord_subtract(Iset, ['L','M','R'], Remaining),
	valid_instruction_set(Remaining).

% The set of valid turns.  Right from N is E, and left from E is N
turns('N', 'E').
turns('E', 'S').
turns('S', 'W').
turns('W', 'N').

% Execute a turn in the indicated direction, returning the new direction
turn(Direction, 'L', NewDirection) :- turns(NewDirection, Direction).
turn(Direction, 'R', NewDirection) :- turns(Direction, NewDirection).

% Effect of movement on position, given the current direction
add('N', in(X,Y), size(_, H), out(X,Yo)) :- Yo is Y + 1, check_grid_pos(Yo, H, "Vertical").
add('S', in(X,Y), size(_, H), out(X,Yo)) :- Yo is Y - 1, check_grid_pos(Yo, H, "Vertical").
add('E', in(X,Y), size(W, _), out(Xo,Y)) :- Xo is X + 1, check_grid_pos(Xo, W, "Horizontal").
add('W', in(X,Y), size(W, _), out(Xo,Y)) :- Xo is X - 1, check_grid_pos(Xo, W, "Horizontal").

% process(pos(X,Y,Facing), size(W, H), [Ins|Instructions], Final)
process(pos(X,Y,Facing), _, [], pos(X,Y,Facing)).
process(pos(X,Y,Facing), Size, ['M'|Instructions], Final) :-
	add(Facing, in(X,Y), Size, out(Xo,Yo)), !, process(pos(Xo,Yo,Facing), Size, Instructions, Final).
process(pos(X,Y,Facing), Size, [T|Instructions], Final) :-
	turn(Facing, T, NewDirection), !, process(pos(X,Y,NewDirection), Size, Instructions, Final).
